File: lib/currency.ts
// 환율 단일 SoT(Single Source of Truth).
// 운영 주석: 이 값은 TS(표시·toUsd)와 Python(scripts/currency.py, 수집·스코어링)이
//   동일한 env 키를 공유해야 일관된다. 두 곳의 env 키를 항상 동기화할 것.
//   클라이언트 번들에서도 동일 환율을 쓰려면 NEXT_PUBLIC_KRW_USD_RATE 가 필요하다
//   (서버 전용 KRW_USD_RATE 는 클라 번들에 주입되지 않으므로 NEXT_PUBLIC_ 을 우선 조회).
const CURRENCY_RATES: Record<string, number> = {
  KRW:
    Number(process.env.NEXT_PUBLIC_KRW_USD_RATE) ||
    Number(process.env.KRW_USD_RATE) ||
    1450,
  JPY: Number(process.env.JPY_USD_RATE) || 150,
  TWD: Number(process.env.TWD_USD_RATE) || 32,
  HKD: Number(process.env.HKD_USD_RATE) || 7.8,
  EUR: Number(process.env.EUR_USD_RATE) || 0.92,
}

const TICKER_SUFFIX_CURRENCY: Record<string, string> = {
  '.KS': 'KRW',
  '.KQ': 'KRW',
  '.T': 'JPY',
  '.TW': 'TWD',
  '.HK': 'HKD',
  '.PA': 'EUR',
}

function getCurrencyRate(ticker: string): number {
  for (const [suffix, currency] of Object.entries(TICKER_SUFFIX_CURRENCY)) {
    if (ticker.endsWith(suffix)) {
      return CURRENCY_RATES[currency]
    }
  }
  return 1
}

export function toUsd(value: number, ticker: string): number {
  return value / getCurrencyRate(ticker)
}

/**
 * KRW→USD 표시 환율(SoT). lib/format.ts 의 formatKrw 등 USD→₩ 역환산 표기에서 사용.
 * toUsd 와 동일한 CURRENCY_RATES.KRW 를 참조하므로 $ 표시와 ₩ 병기가 항상 같은 환율을 쓴다.
 */
export function getKrwRate(): number {
  return CURRENCY_RATES.KRW
}

// ── 표시 통화 토글 SoT ────────────────────────────────────────────────
// 사용자가 화면에서 선택하는 표시 통화. API 응답은 항상 USD 이며, 이 값은
// "어떤 단위로 표기할지"만 결정한다(toUsd 의 데이터 변환과는 별개의 표시 레이어).

/** 화면 표시 통화. 'KRW'=원화(₩), 'USD'=달러($). */
export type Currency = 'KRW' | 'USD'

/** 요구사항: 기본 표기는 원화(₩). */
export const DEFAULT_CURRENCY: Currency = 'KRW'

/**
 * localStorage / inline-script / data-attr 가 공유하는 상수(오타 방지·단일 진실).
 *
 * 주의: app/layout.tsx 의 <head> inline script 는 번들러가 이 상수를 자동
 *   치환하지 못하므로 키('sector-king-currency')·기본값('KRW')을 문자열로
 *   하드코딩한다. 값 변경 시 inline script 와 반드시 동시 수정할 것.
 */
export const CURRENCY_STORAGE_KEY = 'sector-king-currency'

/** <html data-currency="KRW|USD"> 속성명. inline script 와 동기화. */
export const CURRENCY_ATTRIBUTE = 'data-currency'

/** 임의 값이 유효한 Currency 인지 검증(런타임 가드). */
export function isCurrency(v: unknown): v is Currency {
  return v === 'KRW' || v === 'USD'
}

File: scripts/currency.py
#!/usr/bin/env python3
"""통화 정규화 — lib/currency.ts 의 Python 미러.

SoT 일치 (중요):
    이 모듈의 CURRENCY_RATES / TICKER_SUFFIX_CURRENCY 는 `lib/currency.ts` 와 **반드시 동일**해야 한다.
    환율/접미사 추가·변경 시 두 파일을 함께 수정한다(TS 가 1차 SoT, 여기는 Python 미러).
    env 키도 동일하게 사용: KRW_USD_RATE / JPY_USD_RATE / TWD_USD_RATE / HKD_USD_RATE / EUR_USD_RATE.

용도:
    시총·가격성 값을 native 통화 → USD 로 환산해 "정렬·임계 비교" 를 USD 기준으로 강제한다.
    (예: KR 종목 marketCap 은 KRW raw 라 toUsd 누락 시 1450배 큰 값이 항상 위로 올라가 KR 편중 오정렬.)

참고:
    07_market_scope 이후 KRW 외 통화 종목은 제거되지만, SoT 일치를 위해 lib/currency.ts 의
    전체 테이블을 그대로 미러한다(새 거래소 접미사가 다시 추가될 때도 안전).
"""

import os

# lib/currency.ts::CURRENCY_RATES 와 동일 (env fallback 기본값 동일)
CURRENCY_RATES: dict[str, float] = {
    "KRW": float(os.environ.get("KRW_USD_RATE") or 1450),
    "JPY": float(os.environ.get("JPY_USD_RATE") or 150),
    "TWD": float(os.environ.get("TWD_USD_RATE") or 32),
    "HKD": float(os.environ.get("HKD_USD_RATE") or 7.8),
    "EUR": float(os.environ.get("EUR_USD_RATE") or 0.92),
}

# lib/currency.ts::TICKER_SUFFIX_CURRENCY 와 동일
TICKER_SUFFIX_CURRENCY: dict[str, str] = {
    ".KS": "KRW",
    ".KQ": "KRW",
    ".T": "JPY",
    ".TW": "TWD",
    ".HK": "HKD",
    ".PA": "EUR",
}


def get_currency_rate(ticker: str) -> float:
    """티커 접미사로부터 USD 환산 분모(rate)를 반환. 접미사 없으면 1(USD)."""
    for suffix, currency in TICKER_SUFFIX_CURRENCY.items():
        if ticker.endswith(suffix):
            return CURRENCY_RATES[currency]
    return 1.0


def to_usd(value: float | None, ticker: str) -> float | None:
    """native 통화 값을 USD 로 환산. value 가 None 이면 None 반환(불변·안전)."""
    if value is None:
        return None
    return value / get_currency_rate(ticker)
